File: scripts/bible_layout.py
"""Recover verse boundaries from the chapter text this corpus already stores.

The WLC backfill could re-render its chapters from morphhb and diff the result
against the database. LHB and ESV have no such source here — the text in
`core.document_texts` is all there is, so the verses have to be read back out of
it. That is safe only with a guard that fails loudly, so every parser below is
checked for *coverage*: blank out the spans it claims, and whatever is left must
be nothing but verse markers and whitespace. A parser that silently dropped half
a verse would leave that verse's text sitting in a gap, and the check would say
so. Chapters that fail are skipped, never guessed at.

Layouts differ completely between the two editions:

LHB is one verse per block, blocks joined by `VERSE_SEP`, each block
`"{verse} \t{text} "` and the first also carrying the chapter number. It is the
layout the WLC ingest copied, so this parser is the inverse of that renderer.

ESV runs verses together in flowing paragraphs, marks each with `"{n} "`,
breaks poetry across blocks, and puts editorial section headings between verses.
Headings have to be separated from verse text or they land inside a verse's
span, but nothing in the markup distinguishes them: `"David Anointed King"` and
`"Against whom have you raised your voice"` are both untabbed, capitalised and
unpunctuated. So the rule here is deliberately asymmetric — a block is a heading
only when several independent signals agree, and anything short of that stays
part of the verse. Absorbing a heading into a verse makes a span slightly too
wide; promoting a line of Isaiah to a heading would take real scripture out of
the verse that contains it. Only one of those is recoverable.
"""

from __future__ import annotations

import re

VERSE_SEP = "\n\n \n\n"
NBSP = " "

#: A verse marker in the ESV export: the number, then a non-breaking space.
#: Footnote numerals ("waters.1") carry no NBSP, which is what separates them.
ESV_MARK = re.compile(rf"(\d+){NBSP}")

#: What may sit between two spans once the parser has claimed everything it
#: recognises: separators, tabs, and a verse marker. Nothing else.
LHB_GAP = re.compile(r"^\s*(?:\d+ )?\d+ \t\s*$|^\s*$")
#: The double brackets around Mark 16:9-20 and John 7:53-8:11 are the ESV's own
#: text-critical marks. They belong to no verse, so they are allowed to sit in a
#: gap rather than being folded into the reading on either side of them.
ESV_GAP = re.compile(rf"^[\s]*(?:\[\[)?[\s]*(?:\d+{NBSP}[\s]*)?(?:\]\])?[\s]*$")

_SENTENCE_END = tuple(".!?;,:”’\"')-—")


class Region:
    """A claimed stretch of the chapter text."""

    __slots__ = ("kind", "number", "title", "start", "end")

    def __init__(self, kind, start, end, number=None, title=None):
        self.kind, self.start, self.end = kind, start, end
        self.number, self.title = number, title

    def __repr__(self):
        return f"Region({self.kind},{self.number},{self.start}:{self.end})"


def _strip_footnote_digits(s: str) -> str:
    """Drop the footnote numerals the ESV export glues onto a word."""
    return re.sub(r"\d+$", "", s)


def _confident_heading(block: str, following: str | None) -> bool:
    """True only when every signal agrees this block is an editorial heading.

    Four independent tests, all of which a section heading passes and a line of
    verse fails: it is not indented (poetry continuations are), it carries no
    verse marker, it begins a new sentence rather than continuing one, and it
    does not end mid-clause. The fourth test runs after footnote numerals are
    removed, because `"forever;6"` ends in a digit and would otherwise read as
    unpunctuated.

    The last test is the important one: a following block that begins lowercase
    is the rest of a sentence, which means this block was its first half and not
    a heading at all. That single check is what rejects `"Against whom have you
    raised your voice"` / `"and lifted your eyes to the heights?"`.
    """
    stripped = block.strip()
    if not stripped or block.startswith("\t") or ESV_MARK.search(block):
        return False
    if not stripped[0].isupper():
        return False
    if _strip_footnote_digits(stripped).endswith(_SENTENCE_END):
        return False
    tail = following.strip() if following else ""
    return not (tail and tail[0].islower())


def parse_lhb(text: str, chapter: int) -> list[Region] | None:
    """One region per verse, read back out of the renderer's own layout."""
    regions: list[Region] = []
    pos = 0
    for index, block in enumerate(text.split(VERSE_SEP)):
        # The first block normally carries the chapter number as well as the
        # verse ("1 1 \t"). A one-chapter book has none to carry, so Obadiah
        # opens "1 \t" and matches the same pattern as every later block.
        patterns = (
            [rf"^{chapter} (\d+) \t", r"^(\d+) \t"] if index == 0 else [r"^(\d+) \t"]
        )
        match = next((m for m in (re.match(p, block) for p in patterns) if m), None)
        if not match:
            return None
        body = block[match.end() :]
        start = pos + match.end()
        regions.append(
            Region("verse", start, start + len(body.rstrip()), number=int(match.group(1)))
        )
        pos += len(block) + len(VERSE_SEP)
    return regions


def parse_esv(text: str, chapter: int) -> list[Region] | None:
    """Verse regions plus the headings confidently separated from them.

    The first marker in a chapter is the chapter number, not the numeral 1 —
    except in the Psalms, where it introduces an unnumbered superscription and
    verse 1 follows with its own marker. The two cases are told apart by what
    comes next: a following marker of 1 means the first was a superscription.

    Two structural facts do most of the work that guessing at prose cannot.
    Anything before the first verse marker has no verse to belong to and is a
    heading whatever it looks like — which is how the question-form titles
    ("Job Replies: Where Is God?") are caught, since they end in punctuation and
    fail every textual test. And a heading always introduces a verse, so a
    candidate not followed by a marker is a line of verse that merely reads like
    a title; those stay inside their verse, where they can do no harm.
    """
    blocks = text.split(VERSE_SEP)
    offsets, pos = [], 0
    for block in blocks:
        offsets.append(pos)
        pos += len(block) + len(VERSE_SEP)

    marked = [bool(ESV_MARK.search(b)) for b in blocks]
    if not any(marked):
        return None
    opening = marked.index(True)

    heads = {i for i in range(opening) if blocks[i].strip()}
    for i, block in enumerate(blocks):
        if i < opening or marked[i] or not block.strip():
            continue
        nxt = next((j for j in range(i + 1, len(blocks)) if blocks[j].strip()), None)
        # A heading introduces a verse, so the block after it opens with that
        # verse's marker. When the marker sits further in, the text before it is
        # the second half of a verse the heading interrupted, and cutting the
        # verse at the heading would strand it.
        if nxt is None or not re.match(rf"^\t*\d+{NBSP}", blocks[nxt]):
            continue
        if _confident_heading(block, blocks[nxt]):
            heads.add(i)

    marks: list[tuple[int, int]] = []  # (absolute end of marker, number)
    for i, block in enumerate(blocks):
        if i in heads:
            continue
        for m in ESV_MARK.finditer(block):
            marks.append((offsets[i] + m.end(), int(m.group(1))))
    if not marks or marks[0][1] != chapter:
        return None

    # John 8 opens with 7:53, which the ESV prints at the head of the chapter
    # rather than the foot of the previous one. It is a verse, it is citable,
    # and it belongs to the chapter before this one — so it is kept and marked
    # as foreign rather than forcing the whole chapter to be skipped.
    foreign = (
        len(marks) > 2 and marks[1][1] > marks[2][1] and marks[2][1] == 1
    )
    body = marks[2:] if foreign else marks[1:]

    superscription = not foreign and len(marks) > 1 and marks[1][1] == 1
    numbers = [n for _, n in body] if (superscription or foreign) else [1] + [n for _, n in body]
    if numbers != sorted(set(numbers)) or numbers[0] != 1:
        return None

    head_starts = sorted(offsets[h] for h in heads)
    regions: list[Region] = []
    for j, (mark_end, number) in enumerate(marks):
        if j + 1 < len(marks):
            nxt_end, nxt_num = marks[j + 1]
            stop = nxt_end - len(str(nxt_num)) - len(NBSP)
        else:
            stop = len(text)
        for h in head_starts:
            if mark_end < h < stop:
                stop = h
                break
        # Poetry markers are written "\t2\xa0\t", so the verse's own text starts
        # a tab after the marker ends. The span covers the words, not the
        # indentation that lays them out.
        body = text[mark_end:stop]
        begin = mark_end + (len(body) - len(body.lstrip()))
        end = mark_end + len(body.rstrip())
        if begin > end:
            begin = end = mark_end
        if j == 0 and superscription:
            regions.append(Region("superscription", begin, end, title="superscription"))
        elif j == 0 and foreign:
            # The chapter marker itself, carrying only the "[[" that opens the
            # disputed passage. No verse text of its own.
            regions.append(Region("bracket", begin, end))
        elif j == 1 and foreign:
            regions.append(Region("foreign_verse", begin, end, number=number))
        else:
            regions.append(Region("verse", begin, end, number=1 if j == 0 else number))

    for h in sorted(heads):
        block = blocks[h]
        lead = len(block) - len(block.lstrip())
        begin = offsets[h] + lead
        regions.append(Region("heading", begin, begin + len(block.strip()), title=block.strip()))

    regions.sort(key=lambda r: r.start)
    return regions


def coverage_gaps(text: str, regions: list[Region], gap_re: re.Pattern) -> list[str]:
    """Everything the parser did not claim, minus what is allowed to be there.

    This is the guard the whole module rests on. If a parser loses a verse, the
    verse's text turns up here and does not match `gap_re`, so the chapter is
    skipped rather than written with offsets that point at the wrong words.
    """
    bad, cursor = [], 0
    for region in sorted(regions, key=lambda r: r.start):
        if region.start < cursor:
            bad.append(f"overlap at {region.start}")
            continue
        gap = text[cursor : region.start]
        if not gap_re.match(gap):
            bad.append(repr(gap[:60]))
        cursor = region.end
    if not gap_re.match(text[cursor:]):
        bad.append(repr(text[cursor:][:60]))
    return bad


def parse_unmarked(text: str) -> list[Region]:
    """Verses of a chapter that carries no verse numbers at all.

    Leviticus 25 was ingested separately for the dossier work and its verses are
    divided by a bare blank line with nothing to number them. Counting blocks
    from one recovers the numbering only if no verse is missing and none holds a
    blank line of its own, neither of which the text can confirm — so a caller
    must check the result against locators recorded elsewhere before trusting
    it. `backfill_editions` refuses the chapter when it cannot.
    """
    regions, pos = [], 0
    for block in text.split("\n\n"):
        stripped = block.strip()
        if stripped:
            lead = len(block) - len(block.lstrip())
            regions.append(
                Region("verse", pos + lead, pos + lead + len(stripped),
                       number=len(regions) + 1)
            )
        pos += len(block) + 2
    return regions

File: crates/marginalia-parse/src/bin/bible_layout.rs
//! `scripts/bible_layout.py`: recover verse boundaries from stored text.
//!
//! Reads one chapter file, runs the edition parser plus the coverage guard,
//! and prints the regions (and any gaps) as JSON. The layout logic lives
//! here, in the binary — it ran once over a corpus, not per document.

use std::sync::LazyLock;

use marginalia_parse::pystr::{is_lower, is_upper};
use marginalia_text::chars::{is_space, strip};
use marginalia_text::normalize::PY_WS_CLASS;
use regex::Regex;

fn ws() -> String {
    format!("[{PY_WS_CLASS}]")
}

const VERSE_SEP: &str = "\n\n \n\n";
// The literals on the next lines read as a space but are U+00A0:
// the ESV numbers its verses with a non-breaking space.
const NBSP: &str = " ";

static ESV_MARK: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(\d+) ").unwrap());
static LHB_GAP: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(&format!(
        "\\A(?:{w}*(?:\\d+ )?\\d+ \\t{w}*|{w}*)(?:\\n)?\\z",
        w = ws()
    ))
    .unwrap()
});
static ESV_GAP: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(&format!(
        "\\A(?:{w}*(?:\\[\\[)?{w}*(?:\\d+ {w}*)?(?:\\]\\])?{w}*)(?:\\n)?\\z",
        w = ws()
    ))
    .unwrap()
});
static LHB_VERSE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\A(\d+) \t").unwrap());
static FOOTNOTE_DIGITS: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(\d+)((?:\n)?\z)").unwrap());
static ESV_BLOCK_OPEN: LazyLock<Regex> = LazyLock::new(|| Regex::new("\\A\\t*\\d+ ").unwrap());

const SENTENCE_END: [char; 13] = [
    '.', '!', '?', ';', ',', ':', '”', '’', '"', '\'', ')', '-', '—',
];

/// Python string indices are character indices; regex reports bytes. This
/// is the bridge, built once per chapter.
struct CharText {
    chars: Vec<char>,
    byte_to_char: Vec<usize>,
}

impl CharText {
    fn new(text: &str) -> Self {
        let chars: Vec<char> = text.chars().collect();
        let mut byte_to_char = vec![0usize; text.len() + 1];
        for (char_index, (byte_index, ch)) in text.char_indices().enumerate() {
            for slot in &mut byte_to_char[byte_index..byte_index + ch.len_utf8()] {
                *slot = char_index;
            }
        }
        byte_to_char[text.len()] = chars.len();
        Self {
            chars,
            byte_to_char,
        }
    }

    fn slice(&self, start: usize, end: usize) -> String {
        self.chars[start..end].iter().collect()
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct Region {
    kind: String,
    number: Option<u64>,
    title: Option<String>,
    start: usize,
    end: usize,
}

impl Region {
    fn verse(start: usize, end: usize, number: u64) -> Self {
        Region {
            kind: "verse".to_owned(),
            number: Some(number),
            title: None,
            start,
            end,
        }
    }

    fn to_json(&self) -> String {
        let number = self
            .number
            .map(|n| n.to_string())
            .unwrap_or_else(|| "null".to_owned());
        let title = self
            .title
            .as_ref()
            .map_or("null".to_owned(), |t| format!("{t:?}"));
        format!(
            "{{\"kind\":{:?},\"number\":{number},\"title\":{title},\"start\":{},\"end\":{}}}",
            self.kind, self.start, self.end
        )
    }
}

/// Drop the footnote numerals the ESV export glues onto a word.
fn strip_footnote_digits(s: &str) -> String {
    FOOTNOTE_DIGITS.replace(s, "$2").into_owned()
}

/// True only when every signal agrees this block is an editorial heading.
fn confident_heading(block: &str, following: Option<&str>) -> bool {
    let stripped = strip(block);
    if stripped.is_empty() || block.starts_with('\t') || ESV_MARK.is_match(block) {
        return false;
    }
    if !is_upper(stripped.chars().next().unwrap_or('\0')) {
        return false;
    }
    let cleaned = strip_footnote_digits(stripped);
    if cleaned
        .chars()
        .last()
        .is_some_and(|c| SENTENCE_END.contains(&c))
    {
        return false;
    }
    let tail = following.map(|f| strip(f).to_owned()).unwrap_or_default();
    tail.is_empty() || !is_lower(tail.chars().next().unwrap_or('\0'))
}

/// One region per verse, read back out of the renderer's own layout.
fn parse_lhb(text: &str, chart: &CharText, chapter: u64) -> Option<Vec<Region>> {
    let with_chapter = Regex::new(&format!("\\A{chapter} (\\d+) \\t")).unwrap();
    let mut regions = Vec::new();
    let mut pos = 0usize;
    let mut byte_pos = 0usize;
    for (index, block) in text.split(VERSE_SEP).enumerate() {
        let pattern = if index == 0 {
            with_chapter
                .captures(block)
                .and_then(|caps| caps.get(1))
                .or_else(|| LHB_VERSE.captures(block).and_then(|caps| caps.get(1)))?
        } else {
            LHB_VERSE.captures(block)?.get(1)?
        };
        let number: u64 = pattern.as_str().parse().ok()?;
        let match_end = pattern.end();
        // The match ends at the digits; the marker is digits + " \t".
        let body_start = match_end + " \t".len();
        let body: Vec<char> = block[body_start..].chars().collect();
        let trimmed = body.len() - body.iter().rev().take_while(|c| is_space(**c)).count();
        let start = chart.byte_to_char[byte_pos + body_start];
        regions.push(Region::verse(start, start + trimmed, number));
        pos += block.chars().count() + VERSE_SEP.chars().count();
        byte_pos += block.len() + VERSE_SEP.len();
    }
    let _ = pos;
    Some(regions)
}

/// Verse regions plus the headings confidently separated from them.
fn parse_esv(text: &str, chart: &CharText, chapter: u64) -> Option<Vec<Region>> {
    let blocks: Vec<&str> = text.split(VERSE_SEP).collect();
    let mut offsets = Vec::with_capacity(blocks.len());
    let mut pos = 0usize;
    for block in &blocks {
        offsets.push(pos);
        pos += block.chars().count() + VERSE_SEP.chars().count();
    }

    let marked: Vec<bool> = blocks.iter().map(|b| ESV_MARK.is_match(b)).collect();
    if !marked.iter().any(|m| *m) {
        return None;
    }
    let opening = marked.iter().position(|m| *m).unwrap_or(0);

    let mut heads: Vec<usize> = (0..opening)
        .filter(|i| !strip(blocks[*i]).is_empty())
        .collect();
    for (i, block) in blocks.iter().enumerate() {
        if i < opening || marked[i] || strip(block).is_empty() {
            continue;
        }
        let nxt = ((i + 1)..blocks.len()).find(|j| !strip(blocks[*j]).is_empty());
        let Some(nxt) = nxt else { continue };
        if !ESV_BLOCK_OPEN.is_match(blocks[nxt]) {
            continue;
        }
        if confident_heading(block, Some(blocks[nxt])) {
            heads.push(i);
        }
    }

    // (absolute end of marker, number), in document order.
    let mut marks: Vec<(usize, u64)> = Vec::new();
    for (i, block) in blocks.iter().enumerate() {
        if heads.contains(&i) {
            continue;
        }
        let base_byte: usize = blocks[..i].iter().map(|b| b.len() + VERSE_SEP.len()).sum();
        for caps in ESV_MARK.captures_iter(block) {
            let whole = caps.get(0).unwrap_or_else(|| caps.get(1).unwrap());
            let number: u64 = caps.get(1).unwrap().as_str().parse().ok()?;
            marks.push((chart.byte_to_char[base_byte + whole.end()], number));
        }
    }
    if marks.is_empty() || marks[0].1 != chapter {
        return None;
    }

    // John 8 opens with 7:53, printed at the head of the chapter rather
    // than the foot of the previous one: kept, and marked as foreign.
    let foreign = marks.len() > 2 && marks[1].1 > marks[2].1 && marks[2].1 == 1;
    let body = if foreign { &marks[2..] } else { &marks[1..] };

    let superscription = !foreign && marks.len() > 1 && marks[1].1 == 1;
    let numbers: Vec<u64> = if superscription || foreign {
        body.iter().map(|(_, n)| *n).collect()
    } else {
        std::iter::once(1)
            .chain(body.iter().map(|(_, n)| *n))
            .collect()
    };
    let mut sorted = numbers.clone();
    sorted.sort_unstable();
    sorted.dedup();
    if sorted != numbers || numbers.first() != Some(&1) {
        return None;
    }

    let mut head_starts: Vec<usize> = heads.iter().map(|h| offsets[*h]).collect();
    head_starts.sort_unstable();
    let mut regions: Vec<Region> = Vec::new();
    for (j, (mark_end, number)) in marks.iter().enumerate() {
        let mut stop = if j + 1 < marks.len() {
            let (nxt_end, nxt_num) = marks[j + 1];
            nxt_end - nxt_num.to_string().chars().count() - NBSP.chars().count()
        } else {
            chart.chars.len()
        };
        for h in &head_starts {
            if *mark_end < *h && *h < stop {
                stop = *h;
                break;
            }
        }
        // Poetry markers are written "\t2 \t", so the verse's own text
        // starts a tab after the marker ends.
        let body_text = chart.slice(*mark_end, stop);
        let leading = body_text.chars().take_while(|c| is_space(*c)).count();
        let trailing = body_text.chars().rev().take_while(|c| is_space(*c)).count();
        let mut begin = mark_end + leading;
        let mut end =
            mark_end + body_text.chars().count() - trailing.min(body_text.chars().count());
        if begin > end {
            begin = *mark_end;
            end = *mark_end;
        }
        if j == 0 && superscription {
            regions.push(Region {
                kind: "superscription".to_owned(),
                number: None,
                title: Some("superscription".to_owned()),
                start: begin,
                end,
            });
        } else if j == 0 && foreign {
            regions.push(Region {
                kind: "bracket".to_owned(),
                number: None,
                title: None,
                start: begin,
                end,
            });
        } else if j == 1 && foreign {
            regions.push(Region {
                kind: "foreign_verse".to_owned(),
                number: Some(*number),
                title: None,
                start: begin,
                end,
            });
        } else {
            regions.push(Region::verse(begin, end, if j == 0 { 1 } else { *number }));
        }
    }

    for h in heads.iter().copied() {
        let block = blocks[h];
        let lead = block.chars().take_while(|c| is_space(*c)).count();
        let stripped = strip(block);
        regions.push(Region {
            kind: "heading".to_owned(),
            number: None,
            title: Some(stripped.to_owned()),
            start: offsets[h] + lead,
            end: offsets[h] + lead + stripped.chars().count(),
        });
    }

    regions.sort_by_key(|r| r.start);
    Some(regions)
}

/// Verses of a chapter that carries no verse numbers at all.
fn parse_unmarked(text: &str) -> Vec<Region> {
    let mut regions = Vec::new();
    let mut pos = 0usize;
    for block in text.split("\n\n") {
        let stripped = strip(block);
        if !stripped.is_empty() {
            let lead = block.chars().take_while(|c| is_space(*c)).count();
            regions.push(Region::verse(
                pos + lead,
                pos + lead + stripped.chars().count(),
                regions.len() as u64 + 1,
            ));
        }
        pos += block.chars().count() + 2;
    }
    regions
}

/// Everything the parser did not claim, minus what is allowed to be there.
fn coverage_gaps(chart: &CharText, regions: &[Region], gap: &Regex) -> Vec<String> {
    let mut bad = Vec::new();
    let mut cursor = 0usize;
    let mut ordered: Vec<&Region> = regions.iter().collect();
    ordered.sort_by_key(|r| r.start);
    for region in ordered {
        if region.start < cursor {
            bad.push(format!("overlap at {}", region.start));
            continue;
        }
        let gap_text = chart.slice(cursor, region.start);
        if !gap.is_match(&gap_text) {
            bad.push(format!(
                "{:?}",
                gap_text.chars().take(60).collect::<String>()
            ));
        }
        cursor = region.end;
    }
    let tail = chart.slice(cursor, chart.chars.len());
    if !gap.is_match(&tail) {
        bad.push(format!("{:?}", tail.chars().take(60).collect::<String>()));
    }
    bad
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 5 {
        eprintln!("usage: bible_layout <lhb|esv|unmarked> <chapter> <text-file> <gap:lhb|esv>");
        std::process::exit(2);
    }
    let text = std::fs::read_to_string(&args[3]).unwrap_or_else(|err| {
        eprintln!("cannot read {}: {err}", args[3]);
        std::process::exit(1);
    });
    let chapter: u64 = args[2].parse().unwrap_or_else(|_| {
        eprintln!("bad chapter number");
        std::process::exit(2);
    });
    let chart = CharText::new(&text);
    let regions = match args[1].as_str() {
        "lhb" => parse_lhb(&text, &chart, chapter),
        "esv" => parse_esv(&text, &chart, chapter),
        "unmarked" => Some(parse_unmarked(&text)),
        _ => {
            eprintln!("unknown edition");
            std::process::exit(2);
        }
    };
    let gap = if args[4] == "esv" { &ESV_GAP } else { &LHB_GAP };
    match regions {
        None => println!("{{\"regions\":null}}"),
        Some(regions) => {
            let gaps = coverage_gaps(&chart, &regions, gap);
            let body: Vec<String> = regions.iter().map(Region::to_json).collect();
            let gaps_json: Vec<String> = gaps.iter().map(|g| format!("{g:?}")).collect();
            println!(
                "{{\"regions\":[{}],\"gaps\":[{}]}}",
                body.join(","),
                gaps_json.join(",")
            );
        }
    }
}
